### ISSUE TITLE
Depth Images

### ISSUE THREAD

**Issue by nicholscrawford** [huggingface/lerobot#1144]
I'm interested in using LeRobot's dataset for my own work, but it seems like depth images are not yet supported. Is that planned, or in progress? (also are there plans for documentation?)

**Comment by dennisushi** [comment]
Also, segmentation images (e.g., uint16 format), and [point clouds](https://github.com/huggingface/lerobot/issues/1043). For depth images there's [a branch that appears to support them ](https://github.com/huggingface/lerobot/pull/455)

**Comment by nicholscrawford** [comment]
Yeah, I saw the PR, but it seems to be on a much older version. Official/updated support would be appreciated. Seconded on segmentation and point cloud support, that would also be useful.

**Comment by CarolinePascal** [comment]
Hi @nicholscrawford @dennisushi !

Depth images and point clouds are indeed not yet supported in LeRobot datasets. Integrating them is not planned in the near future, but we may consider it if there is enough interest from the community and enough impactful perspectives on the IA model's side.

Best,

Caroline.

**Comment by Tavish9** [comment]
@nicholscrawford you can store the depth images as tensors into parquet. That's the best solution till now.

**Comment by dennisushi** [comment]
I think the issue is mainly with 'compute_stats' which treats any non-image as a state and computers stats for each element - so when you have a datum with multiple channels, the stats are nonsensicall. If that can be addressed, then there's no issue saving the depth or any other input as a tensor?

**Comment by Tavish9** [comment]
see my PR.
- https://github.com/huggingface/lerobot/pull/945

BTW, lerobot is quite exclusive and doesn’t accept PRs from others. 😪

**Comment by CarolinePascal** [comment]
Hi @Tavish9 !

I'm sorry we gave you this exclusivity impression :( We value all contributions from the community, but we've been dealing with huge hardware refactors lately, and didn't get time to properly address hardware related PRs. 

That being said, we'll really try to make up for this once the refactors are merged ! 

Best,

Caroline.

**Comment by Tavish9** [comment]
> Hi [@Tavish9](https://github.com/Tavish9) !
> 
> I'm sorry we gave you this exclusivity impression :( We value all contributions from the community, but we've been dealing with huge hardware refactors lately, and didn't get time to properly address hardware related PRs.
> 
> That being said, we'll really try to make up for this once the refactors are merged !
> 
> Best,
> 
> Caroline.

I understand where lerobot’s work focus lies, but the core of many unresolved issues/prs is the lack of contributors, at least reviewers. 

For small and efficient implementations, they should be quickly accepted and everyone should be encouraged to collaboratively build the community. 
For code that affects the framework’s implementation, it can wait for the main contributor to review before making a decision.

**Comment by CarolinePascal** [comment]
I completely understand and agree with your opinion. Once the refactors are merged, it will be easier to quickly merge PRs, because they will be more relevant in regard to the new codebase. Merging a PR now knowing it will be overwritten in a few days by another fix in the refactor felt a bit inefficient :/

That being said, I'm going to discuss the depth image integration with the team to see what can be done after the cameras refactor #1108 . In the meantime, **feel free to thumbs up 👍 this issue, and spread the word, to show your interest in the feature** !

**Comment by milong26** [comment]
You can modify by yourself, 
1. change depth_map to image in `pyrealsense.py`. (you can no change, this way is to save depth image)
2. mostly in `manipulator.py` you need to add if `camera.type=realsense and use_depth=True`,get color_image and depth_image from read and save them in observation.
3. add new feature in `manipulator.py` like this 
```python
            if type(cam) is IntelRealSenseCamera:
                if cam.use_depth==True:
                    # depth feature
                    key = f"observation.images.{cam_key}Depth"
                    cam_ft[key] = {
                            "shape": (cam.height, cam.width, cam.channels),
                            "names": ["height", "width", "channels"],
                            "info": None,
                        }
```

these can help u save and use depth image (they'll be seemed as color image from another camera, not real depth image, I want the model know it's depth image :( )

Updates: Warning! image will lost accuracy
sadly, the process of encoding-to-video will pollute depth images. I tried to save depth images in 8bit to R channel, 8bit to G channel. From the saved images (I kept them after merging into videos) I can easliy restore the real depth .
But when I try to restore depth from image through video, it couldn't get right value. My working space is small (around 50cm at most) then I found in R channel even small error make large difference.

Maybe `store the depth images as tensors into parquet.` from [Tavish9](https://github.com/Tavish9) is good. Adding some work to read them and store so much data (height*width*16bit for one image, and due to fps we have more, this is just one second :( and due to time more data in one episode )

And I checked the difference between `real image`(saved when recording) and `from video`(got from lerobot datset), I found in some pixels they're same, in some pixels (mostly the edge of depth image) they show difference. (Of course the difference is not large,but it is fatal :< )

**Comment by IrvingF7** [comment]
> Hi [@nicholscrawford](https://github.com/nicholscrawford) [@dennisushi](https://github.com/dennisushi) !
> 
> Depth images and point clouds are indeed not yet supported in LeRobot datasets. Integrating them is not planned in the near future, but we may consider it if there is enough interest from the community and enough impactful perspectives on the IA model's side.
> 
> Best,
> 
> Caroline.

I think support for depth images or 1D images (instead of requiring 3D) would be greatly appreciated. Not only do people store depth as 1D images, but tactile readings are sometimes stored as 1D/greyscale as well.

Technically speaking, it's possible to simply store them as some kind of 1D tensor, but in Lerobot's many model implementations, the image keyword is scanned to decide whether a vision encoder is applied to this data, so the ability to store 1D images as `VISUAL` and `observation.images` is greatly appreciated.

Best

**Comment by milong26** [comment]
@IrvingF7  agree but is 1D image enough? I used to do this but my depth camera can see more than 255 so 8 bit image is not enough.

**Comment by CarolinePascal** [comment]
Hi @milong26, @IrvingF7,

We are currently discussing depth frames integration in [this Discord thread](https://discord.com/channels/1216765309076115607/1398650737172680810). The idea I have in mind so far is to save depth frames as uint16 grayscale PNG files during recording and to encode them using the existing pipeline as grayscale uint10 or uint12 MP4 videos.

It's still WIP for now, so feel free to come up with new ideas !

Best,

Caroline.

**Comment by IrvingF7** [comment]
> [@IrvingF7](https://github.com/IrvingF7) agree but is 1D image enough? I used to do this but my depth camera can see more than 255 so 8 bit image is not enough.

I think this is a good point. I am not familiar with a standardized way to store depth images, so I will leave this to those more knowledgeable about the topic.

**Comment by IrvingF7** [comment]
> Hi [@milong26](https://github.com/milong26), [@IrvingF7](https://github.com/IrvingF7),
> 
> We are currently discussing depth frames integration in [this Discord thread](https://discord.com/channels/1216765309076115607/1398650737172680810). The idea I have in mind so far is to save depth frames as uint16 grayscale PNG files during recording and to encode them using the existing pipeline as grayscale uint10 or uint12 MP4 videos.
> 
> It's still WIP for now, so feel free to come up with new ideas !
> 
> Best,
> 
> Caroline.

Thank you very much for your attention to this. I think what you proposed can be a great first step and certainly would satisfiy what I need at the moment

**Comment by milong26** [comment]
> Hi [@milong26](https://github.com/milong26), [@IrvingF7](https://github.com/IrvingF7),
> 
> We are currently discussing depth frames integration in [this Discord thread](https://discord.com/channels/1216765309076115607/1398650737172680810). The idea I have in mind so far is to save depth frames as uint16 grayscale PNG files during recording and to encode them using the existing pipeline as grayscale uint10 or uint12 MP4 videos.
> 
> It's still WIP for now, so feel free to come up with new ideas !
> 
> Best,
> 
> Caroline.

Thanks for your work!! I tried  before but encoding depth images to vidoes and decoding may destroy depth raw data using default video encoder and decoder. Wish this problem can be solved!

**Comment by muhtar72603** [comment]
> > Hi [@nicholscrawford](https://github.com/nicholscrawford) [@dennisushi](https://github.com/dennisushi) !嗨，[@nicholscrawford](https://github.com/nicholscrawford)[@dennisushi](https://github.com/dennisushi) ！
> > Depth images and point clouds are indeed not yet supported in LeRobot datasets. Integrating them is not planned in the near future, but we may consider it if there is enough interest from the community and enough impactful perspectives on the IA model's side.LeRobot 数据集确实尚不支持深度图像和点云。在不久的将来没有计划整合它们，但如果社区有足够的兴趣和对 IA 模型方面有足够有影响力的观点，我们可能会考虑这样做。
> > Best,  最好
> > Caroline.  卡罗琳。
> 
> I think support for depth images or 1D images (instead of requiring 3D) would be greatly appreciated. Not only do people store depth as 1D images, but tactile readings are sometimes stored as 1D/greyscale as well.我认为对深度图像或 1D 图像（而不是需要 3D）的支持将不胜感激。人们不仅将深度存储为 1D 图像，而且触觉读数有时也存储为 1D/灰度。
> 
> Technically speaking, it's possible to simply store them as some kind of 1D tensor, but in Lerobot's many model implementations, the image keyword is scanned to decide whether a vision encoder is applied to this data, so the ability to store 1D images as `VISUAL` and `observation.images` is greatly appreciated.从技术上讲，可以简单地将它们存储为某种一维张量，但在 Lerobot 的许多模型实现中，扫描图像关键字以决定是否将视觉编码器应用于这些数据，因此将一维图像存储为 `VISUAL` 和 `observation.images` 的能力非常值得赞赏。
> 
> Best  最好

agree!!!!

**Comment by zhanglixuan0720** [comment]
Hi, I’ve been trying to convert my dataset with depth maps into LeRobot’s format, and I wanted to share some of the challenges I’ve encountered:

1. Storing depth maps as image format:
When I convert the depth map to a PIL image and save it, the values read as int16 instead of uint16, and they can’t be visualized properly.

2. Storing depth maps as float format:
Storing the depth map as float causes issues when calculating stats. It seems like the system isn't fully prepared to handle float values for depth images in statistics computations.

3. Storing depth maps as video format:
Converting depth maps to video format results in all values being 1.0 when read, and I can't visualize the depth data properly.

4. Storing depth maps as RGB (3 channels of uint8):
Storing depth maps as RGB, with each channel representing a portion of the depth, works in terms of reading and visualizing, but it’s an imprecise method. However, the depth map looks somewhat colorful when visualized. This method requires additional processing, and I had to introduce a transform process during the data reading step, which adds extra load to the system. 
```python
ef depth2rgb_storage(depth: torch.Tensor) -> torch.Tensor:
    depth = (depth.float() * 10000).to(torch.int32)
    # high 8 
    R = ((depth >> 16) & 0xFF)
    # mid 8
    G = ((depth >> 8) & 0xFF) 
    # low 8
    B = (depth & 0xFF) 
    rgb = torch.stack([R, G, B], dim=-1).to(torch.uint8)
    return rgb


def rgb_storage2depth(rgb: torch.Tensor) -> torch.Tensor:
    rgb = rgb.to(torch.int32)
    R = rgb[..., 0]
    G = rgb[..., 1]
    B = rgb[..., 2]
    depth = (R << 16) | (G << 8) | B
    depth = depth.float() / 10000.0
    return depth
```

I understand that depth maps and point cloud support aren’t fully integrated into LeRobot yet, but it would be extremely valuable to have this feature, especially for projects involving depth sensors.

One improvement that could help would be adding a mechanism to ignore or mask out certain keys during the stats calculation to support arbitrary tensor types (like depth map stored as float32).

**Update: Storing depth maps as RGB (3 channels of uint8) after MP4 conversion**

I found that storing depth maps as RGB and converting them to MP4 leads to significant distortion. While the visual appearance doesn't change much, the depth values are affected, often resulting in errors of two orders of magnitude. This issue likely stems from compression artifacts in the video format, which lose important depth information during encoding and decoding.

**Further Update**:

I modified the encode_video_frames() function in lerobot/datasets/lerobot_dataset.py to use the following parameters:
```python
encode_video_frames(img_dir, temp_path, self.fps, overwrite=True, vcodec='h264', pix_fmt='yuv444p', crf=0, g=0)
```
This adjustment, which uses the h264 codec with a pixel format of yuv444p and a crf value of 0 for lossless compression, effectively reduces the compression distortion. However, there is still a significant depth value error, and the loss of precision persists. It seems that the formats currently supported (such as h264, hevc, and libsvtav1) are not ideal for encoding depth data.

And, storing the depth maps in Parquet format significantly increases the loading overhead.

**Comment by CarolinePascal** [comment]
Hi @zhanglixuan0720,

Thank you for this very precious feedback ! The issues you encountered with video encoding while using RGB depth maps is particularly interesting - I suppose these encoders were optimized for ordinary visual frames and not depth frames... Still, it conforts me in the idea of focusing first on raw depth frames storage and then dig into the video encoding part ! 

Regarding your first points, I think you can investigate the following leads :

1. You need to provide an additional argument to Pillow to save PNGs with `unint16`:
```
img = Image.fromarray(frame, mode="I;16")  # 16-bit grayscale
img.save("frame_16bit.png")               # PNG supports 16-bit
```
And same when you load them. You can draw inspiration from what is explained [here for `float32`](https://huggingface.co/docs/datasets/main/en/depth_estimation#depth-estimation).

2. For now, statistics computation assumes all images have 3 channels, which is no longer the case for depth frames. The [get_feature_stats](https://github.com/huggingface/lerobot/blob/0f551df8f4bad4c504e395ea3df74fc5f714016f/src/lerobot/datasets/compute_stats.py#L425) function should add support for single channel frames !

Best,

Caroline.

**Comment by atyshka** [comment]
@CarolinePascal Are you aware of this project https://github.com/cheind/hue-depth-encoding? Traditional video compression fails on depth data because it is lossy in the luminance while being pretty lossless in the chroma. This project implements a paper by the Realsense team to map depth values to color values and then encode those color images. That GitHub repo has a bunch of experiments on accuracy. I have personally used it for compressed rosbag recording and gotten over 50x compression with very little quality degradation.

**Comment by atyshka** [comment]
I'll be submitting a PR this week for uncompressed UINT16 depth images. Tests are passing, I just need to confirm with real data. After that I'll explore integrating the hue-encoding compression scheme. The only thing I will note with that is that we will probably want to have control over the encoding parameters, in which case it would be good to merge #1744 first.

Also, hue encoding requires a min and max depth range (e.g. 50-5000cm) to map the colors to. We'll need to integrate this metadata with the dataset so that colorized frames can be properly decoded to depth. I'm not sure if we can store arbitrary custom metadata to the info.json but I'll give it a shot.

**Comment by atyshka** [comment]
@CarolinePascal @nicholscrawford Please check out #2604 and let me know if you have any feedback/suggestions!

**Comment by paulpacaud** [comment]
Integrating depth in lerobot is also a major need in our lab. Would be great to support it!

**Comment by Jonetz** [comment]
Adding depth support would be fantastic. In our company, this is currently the biggest barrier to using only LeRobot. There are 16-bit MP4 codecs that can efficiently compress depth maps with minimal loss, keeping storage requirements reasonable while preserving high quality. Are there currently any plans to integrate it in the future?  I saw the is_depth_map flag in the metadata.

**Comment by ReubenLimMonash** [comment]
Adding support for depth maps would also greatly benefit our lab. Hope to see it supported soon!

**Comment by CarolinePascal** [comment]
Hi everyone !

I'm planning on reviewing/merging this [awesome PR](https://github.com/huggingface/lerobot/pull/2604) as soon as I find the time for it. Feel free to start using it in the meantime 🔥 

Best,

Caroline.

**Comment by IrvingF7** [comment]
> Hi everyone !
> 
> I'm planning on reviewing/merging this [awesome PR](https://github.com/huggingface/lerobot/pull/2604) as soon as I find the time for it. Feel free to start using it in the meantime 🔥
> 
> Best,
> 
> Caroline.

awesome! Thank you for your attention to this matter, and appreciate the author who submitted the PR. Could you ping us in this thread when this is released in a future version of lerobot (so we can pip it), thank you!

**Comment by mountainai** [comment]
> I'll be submitting a PR this week for uncompressed UINT16 depth images. Tests are passing, I just need to confirm with real data. After that I'll explore integrating the hue-encoding compression scheme. The only thing I will note with that is that we will probably want to have control over the encoding parameters, in which case it would be good to merge [#1744](https://github.com/huggingface/lerobot/pull/1744) first.
> 
> Also, hue encoding requires a min and max depth range (e.g. 50-5000cm) to map the colors to. We'll need to integrate this metadata with the dataset so that colorized frames can be properly decoded to depth. I'm not sure if we can store arbitrary custom metadata to the info.json but I'll give it a shot.

can you give an example to create a depth-enabled version of dataset via lerobot-record? I do not find any doc to create such a dataset as atyshka/libero-10-depth.

**Comment by whats2000** [comment]
The feature is working on #3253

**Comment by facontidavide** [comment]
I don't want to introduce too much noise into this topic (I apologize in advance) but I was experimenting with lossless compression of depth images and got results that you may find interesting to consider.

https://discourse.openrobotics.org/t/fast-lossless-image-compression-interested/53986/19

May I ask the level of compression you get with this strategy (I guess considerably better than 10% ratio, since there is temporal information)

**Comment by imstevenpmwork** [comment]
Addressed in: https://github.com/huggingface/lerobot/pull/3644
